module/block_iterator.go

```diff
@@ -0,0 +1,114 @@
+	// CreateJob takes a progress reader which is used to read the progress of the iterator
+	// and returns an iterate job that specifies the range of blocks to iterate over
+	CreateJob(IterateProgressReader) (IterateJob, error)
+}
+
+// IterateProgressReader reads the progress of the iterator, useful for resuming the iteration
+// after restart
+type IterateProgressReader interface {
+	// ReadNext reads the next block to iterate
+	// caller must ensure the reader is created by the IterateProgressInitializer,
+	// otherwise ReadNext would return exception.
+	ReadNext() (uint64, error)
+}
+
+// IterateProgressWriter saves the progress of the iterator
+type IterateProgressWriter interface {
+	// SaveNext persists the next block to be iterated
+	SaveNext(uint64) error
+}
+
+// IterateProgressInitializer is an interface for initializing the progress of the iterator
+// a initializer must be used to ensures the initial next block to be iterated is saved in
+// storage before creating the block iterator
+type IterateProgressInitializer interface {
+	Init() (IterateProgressReader, IterateProgressWriter, error)
+}
+
+// BlockIterator is an interface for iterating over blocks
+type BlockIterator interface {
+	// Next returns the next block in the iterator
+	// Note: a block will only be iterated once in a single iteration, however
+	// if the iteration is interrupted (e.g. by a restart), the iterator can be
+	// resumed from the last checkpoint, which might result in the same block being
+	// iterated again.
+	Next() (blockID flow.Identifier, hasNext bool, exception error)
+
+	// Checkpoint saves the current state of the iterator
+	// so that it can be resumed later
+	// when Checkpoint is called, if SaveNextFunc is called with block A,
+	// then after restart, the iterator will resume from A.
+	Checkpoint() error
+}
+
+// IteratorCreator is an interface for creating block iterators
+type IteratorCreator interface {
+	// CreateIterator takes iterate job which specifies the range of blocks to iterate over
+	// and a progress writer which is used to save the progress of the iterator,
+	// and returns a block iterator that can be used to iterate over the blocks
+	// Note: it's up to the implementation to decide how often the progress is saved,
+	// it is wise to consider the trade-off between the performance and the progress saving,
+	// if the progress is saved too often, it might impact the iteration performance, however,
+	// if the progress is only saved at the end of the iteration, then if the iteration
+	// was interrupted, then the iterator will start from the beginning of the range again,
+	// which means some blocks might be iterated multiple times.
+	CreateIterator(IterateJob, IterateProgressWriter) (BlockIterator, error)
+}
+
+type IteratorFactory struct {
+	progressReader IterateProgressReader
+	progressWriter IterateProgressWriter
+	creator        IteratorCreator
+	jobCreator     IteratorJobCreator
+}
+
+func NewIteratorFactory(
```

Comment:
Once the interfaces in the arguments are implemented, then the logic to create the BlockIterator can be reused. That's why, I put this function here along with the interface definitions, so that it's clear to see how the interfaces will be used for creating the block iterator.

Reply:
There will be two NewBlockIterator implementations: NewHeightBasedBlockIterator and NewViewBasedBlockIterator, both of them will need to implement progress initialization and creating job with range of height / view. And these logic are the same for both, so extracting the iteration factory is to reuse them. 

In other words, there will be one IterationFactory, many different BlockIterator creators.